services/app-api/src/resolvers/rate/withdrawRate.ts

```diff
@@ -0,0 +1,103 @@
+    return async (_parent, { input }, context) => {
+        const { user, ctx, tracer } = context
+        const span = tracer?.startSpan('withdrawRate', {}, ctx)
+        setResolverDetailsOnActiveSpan('withdrawRate', user, span)
+
+        const { rateID, updatedReason } = input
+        span?.setAttribute('mcreview.package_id', rateID)
+
+        if (!hasCMSPermissions(user)) {
+            const message = 'user not authorized to withdraw a rate'
+            logError('withdrawRate', message)
+            setErrorAttributesOnActiveSpan(message, span)
+            throw new ForbiddenError(message)
+        }
+
+        const rateWithHistory = await store.findRateWithHistory(rateID)
+
+        if (rateWithHistory instanceof Error) {
+            const errMessage = `Issue finding rate message: ${rateWithHistory.message}`
+            logError('withdrawRate', errMessage)
+            setErrorAttributesOnActiveSpan(errMessage, span)
+
+            if (rateWithHistory instanceof NotFoundError) {
+                throw new GraphQLError(errMessage, {
+                    extensions: {
+                        code: 'NOT_FOUND',
+                        cause: 'DB_ERROR',
+                    },
+                })
+            }
+
+            throw new GraphQLError(errMessage, {
+                extensions: {
+                    code: 'INTERNAL_SERVER_ERROR',
+                    cause: 'DB_ERROR',
+                },
+            })
+        }
+
+        const allowedStatus = ['SUBMITTED', 'RESUBMITTED'].includes(
+            rateWithHistory.consolidatedStatus
+        )
+
+        if (!allowedStatus) {
+            const errMessage = `Attempted to withdraw rate with wrong status: ${rateWithHistory.consolidatedStatus}`
+            logError('withdrawRate', errMessage)
+            setErrorAttributesOnActiveSpan(errMessage, span)
+            throw new UserInputError(errMessage, {
+                argumentName: 'rateID',
+                cause: 'INVALID_PACKAGE_STATUS',
+            })
+        }
+
+        const withdrawnRate = await store.withdrawRate({
+            rateID,
+            updatedByID: user.id,
+            updatedReason,
+        })
+
+        if (withdrawnRate instanceof Error) {
+            const errMessage = `Failed to withdraw rate message: ${withdrawnRate.message}`
+            logError('withdrawRate', errMessage)
+            setErrorAttributesOnActiveSpan(errMessage, span)
+
+            if (withdrawnRate instanceof NotFoundError) {
+                throw new GraphQLError(errMessage, {
+                    extensions: {
+                        code: 'NOT_FOUND',
+                        cause: 'DB_ERROR',
+                    },
+                })
+            }
+
+            throw new GraphQLError(errMessage, {
+                extensions: {
+                    code: 'INTERNAL_SERVER_ERROR',
+                    cause: 'DB_ERROR',
+                },
+            })
+        }
```

Comment:
Could maybe refactor out the error handling, although it's only happening twice so 🤷🏻‍♂️ . Something like:
```typescript
function createRateError(error: Error, context: string = 'withdrawRate'): GraphQLError {
    const errMessage = `${context} message: ${error.message}`
    
    if (error instanceof NotFoundError) {
        return new GraphQLError(errMessage, {
            extensions: {
                code: 'NOT_FOUND',
                cause: 'DB_ERROR',
            },
        })
    }
    
    return new GraphQLError(errMessage, {
        extensions: {
            code: 'INTERNAL_SERVER_ERROR',
            cause: 'DB_ERROR',
        },
    })
}
```

Which would maybe make it easier to read around the error
```typescript
if (rateWithHistory instanceof Error) {
            const error = createRateError(rateWithHistory, 'Fetching rate')
            logError('withdrawRate', error.message)
            setErrorAttributesOnActiveSpan(error.message, span)
            throw error
        }
```

Not a strong opinion though 😂

Reply:
Yea this has been a ongoing goal for us to improve our error handling in the API. We have several tickets in tech maintenance to figure out a way to handle these type of errors.